pkg/controller/routestate/route_state_controller.go

```diff
@@ -0,0 +1,246 @@
+	cachesToSync    []cache.InformerSynced
+	targetNamespace string
+	operatorClient  v1helpers.OperatorClient
+	eventRecorder   events.Recorder
+}
+
+func NewRouteStateController(kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	endpointsGetter corev1client.EndpointsGetter,
+	podsGetter corev1client.PodsGetter,
+	operatorClient v1helpers.OperatorClient,
+	targetNamespace string,
+	recorder events.Recorder,
+) *RouteStateController {
+	c := &RouteStateController{
+		endpointsGetter: endpointsGetter,
+		podGetter:       podsGetter,
+		targetNamespace: targetNamespace,
+		operatorClient:  operatorClient,
+		eventRecorder:   recorder.WithComponentSuffix("route-state-controller"),
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "RouteStateController"),
+	}
+
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Pods().Informer().HasSynced)
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Endpoints().Informer().HasSynced)
+
+	kubeInformersForTargetNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForTargetNamespace.Core().V1().Endpoints().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c *RouteStateController) eventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(routeStateControllerWorkQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(routeStateControllerWorkQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(routeStateControllerWorkQueueKey) },
+	}
+}
+
+// checkPodStatus will check the target pod container status and return list of possible problems.
+func (c *RouteStateController) checkPodStatus(reference *corev1.ObjectReference) []string {
+	pod, err := c.podGetter.Pods(reference.Namespace).Get(reference.Name, metav1.GetOptions{})
+	if err != nil {
+		return []string{fmt.Sprintf("error getting pod %q: %v", reference.Name, err)}
+	}
+	result := []string{}
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		if containerStatus.Ready && containerStatus.State.Running != nil {
+			continue
+		}
+		if terminated := containerStatus.State.Terminated; terminated != nil {
+			result = append(result, fmt.Sprintf("pod %q container %q terminated with %q", pod.Name, containerStatus.Name, terminated.Message))
+		}
+		if containerStatus.RestartCount > 1 {
+			result = append(result, fmt.Sprintf("pod %q container %q restarted %d times", pod.Name, containerStatus.Name, containerStatus.RestartCount))
+		}
+		// TODO: Add more here, like we can detect "pending" pods
+	}
+	return result
+}
+
+// checkEndpointHealthz will check the health of given https://endpointIP:6443/healthz
+// TODO: this is insecure for now, but enough to tell us if we can get through
+func (c *RouteStateController) checkEndpointHealthz(endpointIP string) (bool, error) {
```

Comment:
Only return error, you never return `false, nil` or `true, err` in this function